pkg/storage/internal/actions/funder/change_distribution.go

```diff
@@ -0,0 +1,113 @@
+package funder
+
+import (
+	"crypto/rand"
+	"fmt"
+	"iter"
+	"math/big"
+
+	"github.com/go-softwarelab/common/pkg/seq"
+	"github.com/go-softwarelab/common/pkg/to"
+)
+
+type Randomizer func(max uint64) uint64
+
+type ChangeDistribution struct {
+	initialValue uint64
+	randomizer   Randomizer
+}
+
+func NewChangeDistribution(initialValue uint64, randomizer Randomizer) *ChangeDistribution {
+	return &ChangeDistribution{
+		initialValue: initialValue,
+		randomizer:   randomizer,
+	}
+}
+
+func (d *ChangeDistribution) Distribute(count uint64, amount uint64) iter.Seq[uint64] {
+	if count == 0 || amount == 0 {
+		return seq.Of[uint64]()
+	}
+	if count == 1 {
+		return seq.Of(amount)
+	}
+
+	countSignedInt := int(count) //nolint:gosec // count is always > 1 at this point
+
+	// saturation: a moment when all the outputs are equal to initialValue
+	saturationThreshold := count * d.initialValue
+	if amount > saturationThreshold {
+		base := amount / count
+		remainder := amount % count
+
+		// e.g. For 3 outputs and 10 amount, we have:
+		// base = 3, remainder = 1, then:
+		// distribution = [4, 3, 3]
+		distribution := seq.Concat(
+			seq.Of[uint64](base+remainder),
+			seq.Repeat(base, countSignedInt-1),
+		)
+
+		noise := d.randomNoise(count, distribution)
+
+		var i uint64
+		var v uint64
+		distribution = seq.Map(distribution, func(current uint64) uint64 {
+			// noise[i] - random value for current output (subtraction does not make it less than initialValue)
+			// noise[reverseIndex] - random value subtracted from another output (added to current)
+
+			reverseIndex := count - i - 1
+			v = current - noise[i] + noise[reverseIndex]
+			i++
+			return v
+		})
+
+		return distribution
+	}
+
+	if amount == saturationThreshold {
+		return seq.Repeat(d.initialValue, countSignedInt)
+	}
+
+	// not saturated - at least one output is less than initialValue:
+	for i := uint64(1); i < count; i++ {
+		saturatedOutputs := count - i
```

Comment:
maybe it would be easier to read if you would make a reverse counter:
```suggestion
	for i := count - 1; i >= 1; i-- {
		saturatedOutputs := i
```
of course it could be also simplified to:
```suggestion
	for saturatedOutputs := count - 1; saturatedOutputs >= 1; saturatedOutputs-- {
```